### Pre-processing for Survival Analysis

In [280]:
import os
import numpy as np
import pandas as pd
import geopandas as gpd
import matplotlib.cm as cm
import matplotlib.pyplot as plt
from requests import get
from urllib.parse import urlparse

In [281]:
df = pd.read_csv("Hampsted HS python.csv", low_memory=False)

In [282]:
df.head()

Unnamed: 0      rates_code rates_authority_id rates_authority  \
0        1540  00000131998018          E09000007          Camden   
1        1541  00000131998018          E09000007          Camden   
2        1542  00000131998018          E09000007          Camden   
3        1543  00000131998018          E09000007          Camden   
4        1544  00000131998018          E09000007          Camden   

                region  location_code          address_no  \
0  Inner London - West      102495063  BST & GND FS 55-56   
1  Inner London - West      102495063  BST & GND FS 55-56   
2  Inner London - West      102495063  BST & GND FS 55-56   
3  Inner London - West      102495063  BST & GND FS 55-56   
4  Inner London - West      102495063  BST & GND FS 55-56   

          address_street address_town  pc__pcs  ... V69  V70        V71 V72  \
0  HAMPSTEAD HIGH STREET       LONDON  NW3 1QH  ...   D  NaN  E05000135 NaN   
1  HAMPSTEAD HIGH STREET       LONDON  NW3 1QH  ...   D  NaN  E05000135 NaN   
2  HAMPSTEAD HIGH STREET       LONDON  NW3 1QH  ...   D  NaN  E05000135 NaN   
3  HAMPSTEAD HIGH STREET       LONDON  NW3 1QH  ...   D  NaN  E05000135 NaN   
4  HAMPSTEAD HIGH STREET       LONDON  NW3 1QH  ...   D  NaN  E05000135 NaN   

          V73  V74  V75   V76 V77                                     geometry  
0  2012-03-19    9  CS1  21.0 NaN  POINT (526376.7504090276 185744.3139116995)  
1  2012-03-19    9  CS1  21.0 NaN  POINT (526376.7504090276 185744.3139116995)  
2  2012-03-19    9  CS1  21.0 NaN  POINT (526376.7504090276 185744.3139116995)  
3  2012-03-19    9  CS1  21.0 NaN  POINT (526376.7504090276 185744.3139116995)  
4  2012-03-19    9  CS1  21.0 NaN  POINT (526376.7504090276 185744.3139116995)  

[5 rows x 104 columns]

In [283]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10117 entries, 0 to 10116
Columns: 104 entries, Unnamed: 0 to geometry
dtypes: bool(2), float64(39), int64(12), object(51)
memory usage: 7.9+ MB


In [284]:
df['status_date']

0        2010-04-01
1        2010-04-01
2        2010-04-01
3        2010-04-01
4        2010-04-01
            ...    
10112    2020-03-08
10113    2020-03-08
10114    2020-03-08
10115    2020-11-18
10116    2020-11-18
Name: status_date, Length: 10117, dtype: object

In [285]:
df = df.sort_values(by= ['status_date'])

In [286]:
cols = ['V3', 'V4', 'V5', 'V13', 'V14', 'V15', 'V16', 'V17', 'V19', 'V20', 'V21', 'V26', 'V28', 'V30', 'V36', 'V38', 'V39', 'V48', 'V51', 'V54', 'V55', 'V56', 'V57', 
        'V58', 'V59', 'V60', 'V64', 'V70', 'V72', 'V77']

In [287]:
#deleted those cols columns from the dataset as they only contain nulls

df = df.drop(cols, axis=1)

In [288]:
df['period'].sample(14)

9823    2017-09
2158    2010-12
9399    2018-12
6866    2010-12
2527    2015-03
5225    2018-09
3058    2015-12
4283    2014-03
2087    2015-03
7472    2019-06
1018    2011-12
3441    2012-09
52      2012-06
5487    2018-03
Name: period, dtype: object

In [289]:
from datetime import datetime

In [290]:
df['new_date'] = pd.to_datetime(df['status_date'], errors='coerce')

In [291]:
df['new_date']

7210   2015-01-02
5989   2010-01-04
6638   2010-01-04
5450   2010-01-04
5098   2010-01-04
          ...    
6990   2015-09-28
2986   2017-11-28
8629   2009-10-29
489    2009-10-29
258    2015-11-30
Name: new_date, Length: 10117, dtype: datetime64[ns]

In [292]:
nulls = df['new_date'][df['new_date'].notnull()==False]

In [293]:
nulls

Series([], Name: new_date, dtype: datetime64[ns])

In [294]:
df['new_date'].sample(25)

6780    2010-04-01
2321    2010-04-01
604     2010-04-01
5897    2010-08-17
6579    2010-03-16
431     2013-04-01
9743    2010-04-01
5236    2010-04-01
4085    2019-05-01
1435    2010-04-01
10040   2018-12-12
9586    2015-04-01
9260    2014-02-07
1129    2016-11-14
1842    2017-01-04
6307    2012-04-01
1516    2010-04-01
1953    2011-12-01
5823    2013-12-19
9865    2017-04-01
8789    2010-10-01
9993    2010-04-01
8506    2010-04-01
9711    2010-04-01
6025    2010-04-01
Name: new_date, dtype: datetime64[ns]

In [295]:
df['status_date'].head(5)

7210    01/02/2015
5989    01/04/2010
6638    01/04/2010
5450    01/04/2010
5098    01/04/2010
Name: status_date, dtype: object

In [296]:
df['new_date'].head(5)

7210   2015-01-02
5989   2010-01-04
6638   2010-01-04
5450   2010-01-04
5098   2010-01-04
Name: new_date, dtype: datetime64[ns]

In [297]:
df = df.sort_values(by= ['new_date'])

In [298]:
df['status_date'].head(5)

8722    2009-10-26
8713    2009-10-26
3231    2009-10-26
8720    2009-10-26
3232    2009-10-26
Name: status_date, dtype: object

In [299]:
df['new_date'].head(5)

8722   2009-10-26
8713   2009-10-26
3231   2009-10-26
8720   2009-10-26
3232   2009-10-26
Name: new_date, dtype: datetime64[ns]

In [300]:
old = df[(df['new_date'] < '2010-04-01')]

In [301]:
old

Unnamed: 0      rates_code rates_authority_id rates_authority  \
8722      306446  99121011200013          E09000007          Camden   
8713      306437  99121011200013          E09000007          Camden   
3231       12911  0006901061004A          E09000007          Camden   
8720      306444  99121011200013          E09000007          Camden   
3232       12912  0006901061004A          E09000007          Camden   
...          ...             ...                ...             ...   
6575       24483  00159000610022          E09000007          Camden   
6557       24465  00159000610022          E09000007          Camden   
6569       24477  00159000610022          E09000007          Camden   
6558       24466  00159000610022          E09000007          Camden   
6561       24469     1.59001E+11          E09000007          Camden   

                   region  location_code          address_no address_street  \
8722  Inner London - West     4017622000  BST & GND FLRS 112   HEATH STREET   
8713  Inner London - West     4017622000  BST & GND FLRS 112   HEATH STREET   
3231  Inner London - West     3942586000    BST & GND FS 106   HEATH STREET   
8720  Inner London - West     4017622000  BST & GND FLRS 112   HEATH STREET   
3232  Inner London - West     3942586000    BST & GND FS 106   HEATH STREET   
...                   ...            ...                 ...            ...   
6575  Inner London - West      100499063      BST & GND FS 6     FLASK WALK   
6557  Inner London - West      100499063      BST & GND FS 6     FLASK WALK   
6569  Inner London - West      100499063      BST & GND FS 6     FLASK WALK   
6558  Inner London - West      100499063      BST & GND FS 6     FLASK WALK   
6561  Inner London - West      100499063      BST & GND FS 6     FLASK WALK   

     address_town  pc__pcs  ... V67  V68 V69        V71         V73  V74 V75  \
8722       LONDON  NW3 1DR  ...   S   1G   D  E05000135  2012-03-19    0  CS   
8713       LONDON  NW3 1DR  ...   S   1G   D  E05000135  2012-03-19    0  CS   
3231       LONDON  NW3 1DR  ...   S   1D   D  E05000135  2012-03-19    1  CS   
8720       LONDON  NW3 1DR  ...   S   1G   D  E05000135  2012-03-19    0  CS   
3232       LONDON  NW3 1DR  ...   S   1D   D  E05000135  2012-03-19    1  CS   
...           ...      ...  ...  ..  ...  ..        ...         ...  ...  ..   
6575       LONDON  NW3 1HE  ...   S   2B   D  E05000135  2012-03-19    1  CS   
6557       LONDON  NW3 1HE  ...   S   2B   D  E05000135  2012-03-19    1  CS   
6569       LONDON  NW3 1HE  ...   S   2B   D  E05000135  2012-03-19    1  CS   
6558       LONDON  NW3 1HE  ...   S   2B   D  E05000135  2012-03-19    1  CS   
6561       LONDON  NW3 1HE  ...   S   2B   D  E05000135  2012-03-19    1  CS   

        V76                                     geometry   new_date  
8722  249.0  POINT (526385.7491817896 186065.3075482332) 2009-10-26  
8713  249.0  POINT (526385.7491817896 186065.3075482332) 2009-10-26  
3231  249.0   POINT (526380.749324891 186014.3085337395) 2009-10-26  
8720  249.0  POINT (526385.7491817896 186065.3075482332) 2009-10-26  
3232  249.0   POINT (526380.749324891 186014.3085337395) 2009-10-26  
...     ...                                          ...        ...  
6575  249.0  POINT (526439.7541909092 185749.3106223808) 2010-03-16  
6557  249.0  POINT (526439.7541909092 185749.3106223808) 2010-03-16  
6569  249.0  POINT (526439.7541909092 185749.3106223808) 2010-03-16  
6558  249.0  POINT (526439.7541909092 185749.3106223808) 2010-03-16  
6561  249.0  POINT (526439.7541909092 185749.3106223808) 2010-03-16  

[546 rows x 75 columns]

546 rows of data where the date is less than our study start date and needs changing to 2010-04-01

In [302]:
df['period'].sample(14)

5312    2018-06
5732    2013-06
7384    2019-06
6669    2016-09
4427    2017-03
5272    2019-06
4777    2016-09
1787    2017-03
2508    2010-06
6481    2013-09
8769    2014-03
5088    2017-06
7325    2015-09
9840    2017-12
Name: period, dtype: object

In [303]:
df['new_date'].loc[(df['new_date'] < '2010-04-01')] = '2010-04-01'

/opt/conda/envs/sds2020/lib/python3.7/site-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [304]:
df['new_date'].head(5)

8722    2010-04-01
8713    2010-04-01
3231    2010-04-01
8720    2010-04-01
3232    2010-04-01
Name: new_date, dtype: object

In [305]:
df = df.sort_values(by= ['location_code', 'period'])

In [306]:
df

Unnamed: 0      rates_code rates_authority_id rates_authority  \
7304        25828  00184001B00002          E09000007          Camden   
7305        25829  00184001B00002          E09000007          Camden   
7306        25830  00184001B00002          E09000007          Camden   
7307        25831  00184001B00002          E09000007          Camden   
7308        25832  00184001B00002          E09000007          Camden   
...           ...             ...                ...             ...   
10110      429165  00153007210029          E09000007          Camden   
10111      429166  00153007210029          E09000007          Camden   
10112      429167  00153007210029          E09000007          Camden   
10113      429168  00153007210029          E09000007          Camden   
10114      429169  00153007210029          E09000007          Camden   

                    region  location_code  address_no         address_street  \
7304   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7305   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7306   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7307   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7308   Inner London - West       25638063          1B         DOWNSHIRE HILL   
...                    ...            ...         ...                    ...   
10110  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10111  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10112  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10113  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10114  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   

      address_town  pc__pcs  ...  V67  V68 V69        V71         V73  V74  \
7304        LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0   
7305        LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0   
7306        LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0   
7307        LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0   
7308        LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0   
...            ...      ...  ...  ...  ...  ..        ...         ...  ...   
10110       LONDON  NW3 1QP  ...  NaN  NaN   C  E05000135         NaN    0   
10111       LONDON  NW3 1QP  ...  NaN  NaN   C  E05000135         NaN    0   
10112       LONDON  NW3 1QP  ...  NaN  NaN   C  E05000135         NaN    0   
10113       LONDON  NW3 1QP  ...  NaN  NaN   C  E05000135         NaN    0   
10114       LONDON  NW3 1QP  ...  NaN  NaN   C  E05000135         NaN    0   

      V75    V76                                     geometry  \
7304   MH  437.0  POINT (526838.7620297795 185589.3079796332)   
7305   MH  437.0  POINT (526838.7620297795 185589.3079796332)   
7306   MH  437.0  POINT (526838.7620297795 185589.3079796332)   
7307   MH  437.0  POINT (526838.7620297795 185589.3079796332)   
7308   MH  437.0  POINT (526838.7620297795 185589.3079796332)   
...    ..    ...                                          ...   
10110  CO  203.0  POINT (526450.7546235088 185680.3055356044)   
10111  CO  203.0  POINT (526450.7546235088 185680.3055356044)   
10112  CO  203.0  POINT (526450.7546235088 185680.3055356044)   
10113  CO  203.0  POINT (526450.7546235088 185680.3055356044)   
10114  CO  203.0  POINT (526450.7546235088 185680.3055356044)   

                  new_date  
7304   2010-04-01 00:00:00  
7305   2010-04-01 00:00:00  
7306   2010-04-01 00:00:00  
7307   2010-04-01 00:00:00  
7308   2010-04-01 00:00:00  
...                    ...  
10110  2020-03-08 00:00:00  
10111  2020-03-08 00:00:00  
10112  2020-03-08 00:00:00  
10113  2020-03-08 00:00:00  
10114  2020-03-08 00:00:00  

[10117 rows x 75 columns]

### CHECK FOR VACANT STORES

In [307]:
vacant = df[(df['status'] == False)]

In [308]:
vacant

Unnamed: 0      rates_code rates_authority_id rates_authority  \
7304        25828  00184001B00002          E09000007          Camden   
7305        25829  00184001B00002          E09000007          Camden   
7306        25830  00184001B00002          E09000007          Camden   
7307        25831  00184001B00002          E09000007          Camden   
7308        25832  00184001B00002          E09000007          Camden   
...           ...             ...                ...             ...   
10093      419528  0029500482000A          E09000007          Camden   
10094      419529  0029500482000A          E09000007          Camden   
10095      419530  0029500482000A          E09000007          Camden   
10096      419531  0029500482000A          E09000007          Camden   
10097      419532  0029500482000A          E09000007          Camden   

                    region  location_code address_no  address_street  \
7304   Inner London - West       25638063         1B  DOWNSHIRE HILL   
7305   Inner London - West       25638063         1B  DOWNSHIRE HILL   
7306   Inner London - West       25638063         1B  DOWNSHIRE HILL   
7307   Inner London - West       25638063         1B  DOWNSHIRE HILL   
7308   Inner London - West       25638063         1B  DOWNSHIRE HILL   
...                    ...            ...        ...             ...   
10093  Inner London - West    11765732000         48    ROSSLYN HILL   
10094  Inner London - West    11765732000         48    ROSSLYN HILL   
10095  Inner London - West    11765732000         48    ROSSLYN HILL   
10096  Inner London - West    11765732000         48    ROSSLYN HILL   
10097  Inner London - West    11765732000         48    ROSSLYN HILL   

      address_town  pc__pcs  ... V67  V68 V69        V71         V73  V74 V75  \
7304        LONDON  NW3 1NR  ...   S   2G   D  E05000135  2012-03-19    0  MH   
7305        LONDON  NW3 1NR  ...   S   2G   D  E05000135  2012-03-19    0  MH   
7306        LONDON  NW3 1NR  ...   S   2G   D  E05000135  2012-03-19    0  MH   
7307        LONDON  NW3 1NR  ...   S   2G   D  E05000135  2012-03-19    0  MH   
7308        LONDON  NW3 1NR  ...   S   2G   D  E05000135  2012-03-19    0  MH   
...            ...      ...  ...  ..  ...  ..        ...         ...  ...  ..   
10093       LONDON  NW3 1NH  ...   S   1J   D  E05000135  2012-03-19    1  CR   
10094       LONDON  NW3 1NH  ...   S   1J   D  E05000135  2012-03-19    1  CR   
10095       LONDON  NW3 1NH  ...   S   1J   D  E05000135  2012-03-19    1  CR   
10096       LONDON  NW3 1NH  ...   S   1J   D  E05000135  2012-03-19    1  CR   
10097       LONDON  NW3 1NH  ...   S   1J   D  E05000135  2012-03-19    1  CR   

         V76                                     geometry             new_date  
7304   437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
7305   437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
7306   437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
7307   437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
7308   437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
...      ...                                          ...                  ...  
10093  234.0  POINT (526753.7573694737 185603.3076923557)  2019-07-16 00:00:00  
10094  234.0  POINT (526753.7573694737 185603.3076923557)  2019-07-16 00:00:00  
10095  234.0  POINT (526753.7573694737 185603.3076923557)  2019-07-16 00:00:00  
10096  234.0  POINT (526753.7573694737 185603.3076923557)  2019-07-16 00:00:00  
10097  234.0  POINT (526753.7573694737 185603.3076923557)  2019-07-16 00:00:00  

[610 rows x 75 columns]

In [309]:
# 610 vacant stores but need to establish if any were vacant prior to study period 

## LEFT CENSOR 

In [310]:
# do i need to make the location codes keys?

In [311]:
remove = vacant[(vacant['status_date'] == '2010-04-01')]

In [312]:
remove

Unnamed: 0      rates_code rates_authority_id rates_authority  \
7304       25828  00184001B00002          E09000007          Camden   
7305       25829  00184001B00002          E09000007          Camden   
7306       25830  00184001B00002          E09000007          Camden   
7307       25831  00184001B00002          E09000007          Camden   
7308       25832  00184001B00002          E09000007          Camden   
...          ...             ...                ...             ...   
3515       21027  00148000100036          E09000007          Camden   
3516       21028  00148000100036          E09000007          Camden   
3517       21029  00148000100036          E09000007          Camden   
3518       21030  00148000100036          E09000007          Camden   
3519       21031  00148000100036          E09000007          Camden   

                   region  location_code address_no  address_street  \
7304  Inner London - West       25638063         1B  DOWNSHIRE HILL   
7305  Inner London - West       25638063         1B  DOWNSHIRE HILL   
7306  Inner London - West       25638063         1B  DOWNSHIRE HILL   
7307  Inner London - West       25638063         1B  DOWNSHIRE HILL   
7308  Inner London - West       25638063         1B  DOWNSHIRE HILL   
...                   ...            ...        ...             ...   
3515  Inner London - West     5729723000  3RD FLR 1      HOLLY HILL   
3516  Inner London - West     5729723000  3RD FLR 1      HOLLY HILL   
3517  Inner London - West     5729723000  3RD FLR 1      HOLLY HILL   
3518  Inner London - West     5729723000  3RD FLR 1      HOLLY HILL   
3519  Inner London - West     5729723000  3RD FLR 1      HOLLY HILL   

     address_town  pc__pcs  ...  V67  V68 V69        V71         V73  V74 V75  \
7304       LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0  MH   
7305       LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0  MH   
7306       LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0  MH   
7307       LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0  MH   
7308       LONDON  NW3 1NR  ...    S   2G   D  E05000135  2012-03-19    0  MH   
...           ...      ...  ...  ...  ...  ..        ...         ...  ...  ..   
3515       LONDON  NW3 6UB  ...  NaN  NaN   C  E05000133         NaN    0  CO   
3516       LONDON  NW3 6UB  ...  NaN  NaN   C  E05000133         NaN    0  CO   
3517       LONDON  NW3 6UB  ...  NaN  NaN   C  E05000133         NaN    0  CO   
3518       LONDON  NW3 6UB  ...  NaN  NaN   C  E05000133         NaN    0  CO   
3519       LONDON  NW3 6UB  ...  NaN  NaN   C  E05000133         NaN    0  CO   

        V76                                     geometry             new_date  
7304  437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
7305  437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
7306  437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
7307  437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
7308  437.0  POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00  
...     ...                                          ...                  ...  
3515  203.0    POINT (526353.754357001 185775.309671689)  2010-04-01 00:00:00  
3516  203.0    POINT (526353.754357001 185775.309671689)  2010-04-01 00:00:00  
3517  203.0    POINT (526353.754357001 185775.309671689)  2010-04-01 00:00:00  
3518  203.0    POINT (526353.754357001 185775.309671689)  2010-04-01 00:00:00  
3519  203.0    POINT (526353.754357001 185775.309671689)  2010-04-01 00:00:00  

[125 rows x 75 columns]

In [313]:
df['locaton_code'] = df['location_code'].astype("category")

In [314]:
df['location_code'].unique()

array([   25638063,    32905063,    32907063,    32912063,    32945063,
          32948063,    41376063,    41380063,    41404063,    41419063,
          41422063,    41424063,    41429063,    41440063,    41452063,
          41465063,    41468063,    41472063,    41481063,    44969063,
          44977063,    44984063,    44990063,    44992063,    45002063,
          45010063,    45060063,    45067063,    45075063,    45077063,
          45082063,    45099063,    45111063,    45113063,    45118063,
          45119063,    46955063,    47002063,    67617063,    67618063,
          74478063,    95061063,    95068063,    95134063,    95135063,
          98036063,    98039063,    98040063,    98041063,    98043063,
          98048063,    98049063,    98050063,    98051063,    98052063,
          98053063,    98054063,    98056063,    98057063,   100496063,
         100498063,   100499063,   100501063,   100790063,   102339063,
         102342063,   102345063,   102353063,   102354063,   102

In [315]:
df['location_code'].value_counts()

220547063      44
2482984000     44
100790063      44
3942586000     44
176189063      44
               ..
11929598000    13
12229537000     7
13174048000     5
13174051000     5
13026783000     2
Name: location_code, Length: 261, dtype: int64

In [316]:
 # 261 unique locations in Hampstead HighStreet

In [317]:
locations = df['location_code'].unique()

In [318]:
df.loc[df['name'].isnull(), 'name'] = df['location_code']

In [319]:
df

Unnamed: 0      rates_code rates_authority_id rates_authority  \
7304        25828  00184001B00002          E09000007          Camden   
7305        25829  00184001B00002          E09000007          Camden   
7306        25830  00184001B00002          E09000007          Camden   
7307        25831  00184001B00002          E09000007          Camden   
7308        25832  00184001B00002          E09000007          Camden   
...           ...             ...                ...             ...   
10110      429165  00153007210029          E09000007          Camden   
10111      429166  00153007210029          E09000007          Camden   
10112      429167  00153007210029          E09000007          Camden   
10113      429168  00153007210029          E09000007          Camden   
10114      429169  00153007210029          E09000007          Camden   

                    region  location_code  address_no         address_street  \
7304   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7305   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7306   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7307   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7308   Inner London - West       25638063          1B         DOWNSHIRE HILL   
...                    ...            ...         ...                    ...   
10110  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10111  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10112  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10113  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10114  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   

      address_town  pc__pcs  ...  V68  V69        V71         V73  V74  V75  \
7304        LONDON  NW3 1NR  ...   2G    D  E05000135  2012-03-19    0   MH   
7305        LONDON  NW3 1NR  ...   2G    D  E05000135  2012-03-19    0   MH   
7306        LONDON  NW3 1NR  ...   2G    D  E05000135  2012-03-19    0   MH   
7307        LONDON  NW3 1NR  ...   2G    D  E05000135  2012-03-19    0   MH   
7308        LONDON  NW3 1NR  ...   2G    D  E05000135  2012-03-19    0   MH   
...            ...      ...  ...  ...  ...        ...         ...  ...  ...   
10110       LONDON  NW3 1QP  ...  NaN    C  E05000135         NaN    0   CO   
10111       LONDON  NW3 1QP  ...  NaN    C  E05000135         NaN    0   CO   
10112       LONDON  NW3 1QP  ...  NaN    C  E05000135         NaN    0   CO   
10113       LONDON  NW3 1QP  ...  NaN    C  E05000135         NaN    0   CO   
10114       LONDON  NW3 1QP  ...  NaN    C  E05000135         NaN    0   CO   

         V76                                     geometry  \
7304   437.0  POINT (526838.7620297795 185589.3079796332)   
7305   437.0  POINT (526838.7620297795 185589.3079796332)   
7306   437.0  POINT (526838.7620297795 185589.3079796332)   
7307   437.0  POINT (526838.7620297795 185589.3079796332)   
7308   437.0  POINT (526838.7620297795 185589.3079796332)   
...      ...                                          ...   
10110  203.0  POINT (526450.7546235088 185680.3055356044)   
10111  203.0  POINT (526450.7546235088 185680.3055356044)   
10112  203.0  POINT (526450.7546235088 185680.3055356044)   
10113  203.0  POINT (526450.7546235088 185680.3055356044)   
10114  203.0  POINT (526450.7546235088 185680.3055356044)   

                  new_date  locaton_code  
7304   2010-04-01 00:00:00      25638063  
7305   2010-04-01 00:00:00      25638063  
7306   2010-04-01 00:00:00      25638063  
7307   2010-04-01 00:00:00      25638063  
7308   2010-04-01 00:00:00      25638063  
...                    ...           ...  
10110  2020-03-08 00:00:00   13174051000  
10111  2020-03-08 00:00:00   13174051000  
10112  2020-03-08 00:00:00   13174051000  
10113  2020-03-08 00:00:00   13174051000  
10114  2020-03-08 00:00:00

In [320]:
df['status_date'] = pd.to_datetime(df['status_date'], errors='coerce')

In [321]:
df['status_date'].iloc[df['status_date'] < '2010-04-01'] = '04/01/2010'

/opt/conda/envs/sds2020/lib/python3.7/site-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [322]:
df['status_date'] = pd.to_datetime(df['status_date'], errors='coerce')

In [323]:
df.sample(15)

Unnamed: 0      rates_code rates_authority_id rates_authority  \
1946       11626  00069005730048          E09000007          Camden   
7319       25843  00184001B00002          E09000007          Camden   
3459       20971  00148000100007          E09000007          Camden   
8435      306159  99121002000014          E09000007          Camden   
6448       24356  00159000210014          E09000007          Camden   
5128       22640  00153004700013          E09000007          Camden   
7271       25795  00184000201115          E09000007          Camden   
3344       13024  00076002930274          E09000007          Camden   
7628       43485  00295003610027          E09000007          Camden   
5352       22864  00153005910019          E09000007          Camden   
4160       21672  00153000910018          E09000007          Camden   
4289       21801  00153001200003          E09000007          Camden   
9950      399686  00295002600002          E09000007          Camden   
9941      399213  00153006700904          E09000007          Camden   
1661       11341  00069004510006          E09000007          Camden   

                   region  location_code                   address_no  \
1946  Inner London - West     5754800000    GND F & 1ST F FRONT 57-61   
7319  Inner London - West       25638063                           1B   
3459  Inner London - West      206403063                            1   
8435  Inner London - West     5168955000              BST & GND FS 20   
6448  Inner London - West      100498063               BST & GND FS 2   
5128  Inner London - West     3645904000                   GND FLR 47   
7271  Inner London - West      218768063         CAR SPACES 1-11 AT 2   
3344  Inner London - West      105131063                     GND F 27   
7628  Inner London - West       98040063                     1ST F 36   
5352  Inner London - West      102499063              BST & GND FS 59   
4160  Inner London - West      102339063               BST & GND FS 9   
4289  Inner London - West      102342063              BST & GND FS 12   
9950  Inner London - West     6933846000  HAMPSTEAD POLICE STATION 26   
9941  Inner London - West     5044931000  VODAFONE SITE (48834) AT 67   
1661  Inner London - West      176189063               GND F & BST 45   

             address_street address_town  pc__pcs  ...  V68  V69        V71  \
1946           HEATH STREET       LONDON  NW3 6UG  ...   2H    D  E05000135   
7319         DOWNSHIRE HILL       LONDON  NW3 1NR  ...   2G    D  E05000135   
3459             HOLLY HILL       LONDON  NW3 6UB  ...   1S    D  E05000133   
8435           HEATH STREET       LONDON  NW3 6TE  ...   3B    D  E05000135   
6448             FLASK WALK       LONDON  NW3 1HE  ...   1G    D  E05000135   
5128  HAMPSTEAD HIGH STREET       LONDON  NW3 1QG  ...   1U    D  E05000135   
7271         DOWNSHIRE HILL       LONDON  NW3 1NR  ...  NaN    N  E05000135   
3344                NEW END       LONDON  NW3 1JE  ...  NaN    C  E05000135   
7628           ROSSLYN HILL       LONDON  NW3 1NH  ...   1X    D  E05000135   
5352  HAMPSTEAD HIGH STREET       LONDON  NW3 1QH  ...   1D    D  E05000135   
4160  HAMPSTEAD HIGH STREET       LONDON  NW3 1PR  ...   1R    D  E05000135   
4289  HAMPSTEAD HIGH STREET       LONDON  NW3 1PX  ...   1A    D  E05000135   
9950           ROSSLYN HILL       LONDON  NW3 1PD  ...   1P    D  E05000135   
9941  HAMPSTEAD HIGH STREET       LONDON  NW3 1QP  ...  NaN    N  E05000135   
1661           HEATH STREET       LONDON  NW3 6UA  ...   1P    D  E05000133   

             V73  V74  V75    V76  \
1946  2012-03-19    5   CS  249.0   
7319  2012-03-19    0   MH  437.0   
3459  2012-03-19    3   CO  203.0   
8435  2012-03-19    3   CS  249.0   
6448  2012-03-19    0   CS  249.0   
5128  2012-03-19    3   CS  249.0   
7271         NaN    5  CP1   43.0   
3344         NaN    0   CS  249.0   
7628  2012-03-19    0   MH  437.0   
5352  2012-03-19    0   CS  249.0   
4160  2012-03-19    0   CS  

### change period to a date format eg. 2021-03 = 01/04/2021

In [324]:
df['period2'] = pd.to_datetime(df['period'])

In [325]:
df['period2'] = df['period2'] + pd.DateOffset(months=1)

In [326]:
df['duration'] = df['period2'] - df['status_date']

In [327]:
df['duration'] = df['duration'].dt.days

In [328]:
df

Unnamed: 0      rates_code rates_authority_id rates_authority  \
7304        25828  00184001B00002          E09000007          Camden   
7305        25829  00184001B00002          E09000007          Camden   
7306        25830  00184001B00002          E09000007          Camden   
7307        25831  00184001B00002          E09000007          Camden   
7308        25832  00184001B00002          E09000007          Camden   
...           ...             ...                ...             ...   
10110      429165  00153007210029          E09000007          Camden   
10111      429166  00153007210029          E09000007          Camden   
10112      429167  00153007210029          E09000007          Camden   
10113      429168  00153007210029          E09000007          Camden   
10114      429169  00153007210029          E09000007          Camden   

                    region  location_code  address_no         address_street  \
7304   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7305   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7306   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7307   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7308   Inner London - West       25638063          1B         DOWNSHIRE HILL   
...                    ...            ...         ...                    ...   
10110  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10111  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10112  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10113  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10114  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   

      address_town  pc__pcs  ...        V71         V73 V74 V75    V76  \
7304        LONDON  NW3 1NR  ...  E05000135  2012-03-19   0  MH  437.0   
7305        LONDON  NW3 1NR  ...  E05000135  2012-03-19   0  MH  437.0   
7306        LONDON  NW3 1NR  ...  E05000135  2012-03-19   0  MH  437.0   
7307        LONDON  NW3 1NR  ...  E05000135  2012-03-19   0  MH  437.0   
7308        LONDON  NW3 1NR  ...  E05000135  2012-03-19   0  MH  437.0   
...            ...      ...  ...        ...         ...  ..  ..    ...   
10110       LONDON  NW3 1QP  ...  E05000135         NaN   0  CO  203.0   
10111       LONDON  NW3 1QP  ...  E05000135         NaN   0  CO  203.0   
10112       LONDON  NW3 1QP  ...  E05000135         NaN   0  CO  203.0   
10113       LONDON  NW3 1QP  ...  E05000135         NaN   0  CO  203.0   
10114       LONDON  NW3 1QP  ...  E05000135         NaN   0  CO  203.0   

                                          geometry             new_date  \
7304   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
7305   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
7306   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
7307   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
7308   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
...                                            ...                  ...   
10110  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10111  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10112  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10113  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10114  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   

       locaton_code    period2  duration  
7304       25638063 2010-07-01        91  
7305       25638063 2010-10-01       183  
7306       25638063 2011-01-01       275  
7307       25638063 2011-04-01       365  
7308       25638063 2011-07-01       456  
...             ...        ...       ...  
10110   13174051000 2020-04-01        24  
10111   13174051000 2020-07-01       115  
1011

In [329]:
df['study-start'] = '04/01/2010'

In [330]:
df['study-start'] = pd.to_datetime(df['study-start'], errors='coerce')

In [331]:
df['time-of-entry'] = df['status_date'] -  df['study-start']

In [332]:
df['time-at-vacant'] = 0

In [333]:
df['vacant'] = 0

In [334]:
df

Unnamed: 0      rates_code rates_authority_id rates_authority  \
7304        25828  00184001B00002          E09000007          Camden   
7305        25829  00184001B00002          E09000007          Camden   
7306        25830  00184001B00002          E09000007          Camden   
7307        25831  00184001B00002          E09000007          Camden   
7308        25832  00184001B00002          E09000007          Camden   
...           ...             ...                ...             ...   
10110      429165  00153007210029          E09000007          Camden   
10111      429166  00153007210029          E09000007          Camden   
10112      429167  00153007210029          E09000007          Camden   
10113      429168  00153007210029          E09000007          Camden   
10114      429169  00153007210029          E09000007          Camden   

                    region  location_code  address_no         address_street  \
7304   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7305   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7306   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7307   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7308   Inner London - West       25638063          1B         DOWNSHIRE HILL   
...                    ...            ...         ...                    ...   
10110  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10111  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10112  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10113  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10114  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   

      address_town  pc__pcs  ...    V76  \
7304        LONDON  NW3 1NR  ...  437.0   
7305        LONDON  NW3 1NR  ...  437.0   
7306        LONDON  NW3 1NR  ...  437.0   
7307        LONDON  NW3 1NR  ...  437.0   
7308        LONDON  NW3 1NR  ...  437.0   
...            ...      ...  ...    ...   
10110       LONDON  NW3 1QP  ...  203.0   
10111       LONDON  NW3 1QP  ...  203.0   
10112       LONDON  NW3 1QP  ...  203.0   
10113       LONDON  NW3 1QP  ...  203.0   
10114       LONDON  NW3 1QP  ...  203.0   

                                          geometry             new_date  \
7304   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
7305   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
7306   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
7307   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
7308   POINT (526838.7620297795 185589.3079796332)  2010-04-01 00:00:00   
...                                            ...                  ...   
10110  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10111  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10112  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10113  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10114  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   

      locaton_code    period2  duration study-start  time-of-entry  \
7304      25638063 2010-07-01        91  2010-04-01         0 days   
7305      25638063 2010-10-01       183  2010-04-01         0 days   
7306      25638063 2011-01-01       275  2010-04-01         0 days   
7307      25638063 2011-04-01       365  2010-04-01         0 days   
7308      25638063 2011-07-01       456  2010-04-01         0 days   
...            ...        ...       ...         ...            ...   
10110  13174051000 2020-04-01        24  2010-04-01      3629 days   
10111  13174051000 2020-07-01       115  2010-04-01      3629 days   
10112  13174051000 2020-10-01       207  2010-04-01      3629 days   
10113  13174051000 2021-01-01       299  2010-04-01      3629 d

In [335]:
drop = df[(df['status'] == False) & (df['status_date'] == '2010-04-01')]

In [336]:
drop.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 134 entries, 7304 to 3519
Data columns (total 82 columns):
 #   Column                Non-Null Count  Dtype          
---  ------                --------------  -----          
 0   Unnamed: 0            134 non-null    int64          
 1   rates_code            134 non-null    object         
 2   rates_authority_id    134 non-null    object         
 3   rates_authority       134 non-null    object         
 4   region                134 non-null    object         
 5   location_code         134 non-null    int64          
 6   address_no            134 non-null    object         
 7   address_street        134 non-null    object         
 8   address_town          134 non-null    object         
 9   pc__pcs               134 non-null    object         
 10  use_category          134 non-null    object         
 11  scat_code             134 non-null    int64          
 12  sub_category          134 non-null    object         
 13  c

In [337]:
# 134 rows will be dropped due to being vacant before the study started

In [338]:
df = df.drop(df[(df['status'] == False) & (df['status_date'] == '2010-04-01')].index)

In [339]:
df

Unnamed: 0      rates_code rates_authority_id rates_authority  \
7309        25833  00184001B00002          E09000007          Camden   
7310        25834  00184001B00002          E09000007          Camden   
7311        25835  00184001B00002          E09000007          Camden   
7312        25836  00184001B00002          E09000007          Camden   
7313        25837  00184001B00002          E09000007          Camden   
...           ...             ...                ...             ...   
10110      429165  00153007210029          E09000007          Camden   
10111      429166  00153007210029          E09000007          Camden   
10112      429167  00153007210029          E09000007          Camden   
10113      429168  00153007210029          E09000007          Camden   
10114      429169  00153007210029          E09000007          Camden   

                    region  location_code  address_no         address_street  \
7309   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7310   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7311   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7312   Inner London - West       25638063          1B         DOWNSHIRE HILL   
7313   Inner London - West       25638063          1B         DOWNSHIRE HILL   
...                    ...            ...         ...                    ...   
10110  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10111  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10112  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10113  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   
10114  Inner London - West    13174051000  1ST FLR 72  HAMPSTEAD HIGH STREET   

      address_town  pc__pcs  ...    V76  \
7309        LONDON  NW3 1NR  ...  437.0   
7310        LONDON  NW3 1NR  ...  437.0   
7311        LONDON  NW3 1NR  ...  437.0   
7312        LONDON  NW3 1NR  ...  437.0   
7313        LONDON  NW3 1NR  ...  437.0   
...            ...      ...  ...    ...   
10110       LONDON  NW3 1QP  ...  203.0   
10111       LONDON  NW3 1QP  ...  203.0   
10112       LONDON  NW3 1QP  ...  203.0   
10113       LONDON  NW3 1QP  ...  203.0   
10114       LONDON  NW3 1QP  ...  203.0   

                                          geometry             new_date  \
7309   POINT (526838.7620297795 185589.3079796332)  2011-08-08 00:00:00   
7310   POINT (526838.7620297795 185589.3079796332)  2011-08-08 00:00:00   
7311   POINT (526838.7620297795 185589.3079796332)  2011-08-08 00:00:00   
7312   POINT (526838.7620297795 185589.3079796332)  2011-08-08 00:00:00   
7313   POINT (526838.7620297795 185589.3079796332)  2011-08-08 00:00:00   
...                                            ...                  ...   
10110  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10111  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10112  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10113  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   
10114  POINT (526450.7546235088 185680.3055356044)  2020-03-08 00:00:00   

      locaton_code    period2  duration study-start  time-of-entry  \
7309      25638063 2011-10-01        54  2010-04-01       494 days   
7310      25638063 2012-01-01       146  2010-04-01       494 days   
7311      25638063 2012-04-01       237  2010-04-01       494 days   
7312      25638063 2012-07-01       328  2010-04-01       494 days   
7313      25638063 2012-10-01       420  2010-04-01       494 days   
...            ...        ...       ...         ...            ...   
10110  13174051000 2020-04-01        24  2010-04-01      3629 days   
10111  13174051000 2020-07-01       115  2010-04-01      3629 days   
10112  13174051000 2020-10-01       207  2010-04-01      3629 days   
10113  13174051000 2021-01-01       299  2010-04-01      3629 d

In [340]:
# 10117 - 134 = 9983 rows left

In [341]:
df['location_code'] = df['location_code'].astype("category")

In [342]:
locations = list(df['location_code'].unique())

In [343]:
df = df.set_index('location_code')

In [344]:
df['survival'] = 0

In [345]:
df['survival'] = df['duration']

In [346]:
df['entry'] = df['time-of-entry'].dt.days

In [347]:
df['time-at-vacant'] = df['survival'] + df['entry']

In [348]:
df['stat'] = df['status'].astype(int)

In [349]:
df["StatusChange"] =  df['stat'].shift() != df['stat']

In [350]:
df["StatusChange"].iloc[0] = False

/opt/conda/envs/sds2020/lib/python3.7/site-packages/pandas/core/indexing.py:670: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  iloc._setitem_with_indexer(indexer, value)


In [351]:
df["TenantChange"] =  df['name'].shift(-1) != df['name'] 

In [352]:
df["TenantChange"].iloc[0] = False

In [353]:
subset = df[(df['StatusChange'] == False)  & (df['TenantChange'] == True) & (df['status'] == True)]

In [354]:
subset

Unnamed: 0       rates_code rates_authority_id rates_authority  \
location_code                                                                   
25638063            25871   00184001B00002          E09000007          Camden   
32905063            24639   00159001009992          E09000007          Camden   
32907063            24701   00159001200006          E09000007          Camden   
32907063            24727   00159001200006          E09000007          Camden   
32912063            24389   00159000410016          E09000007          Camden   
...                   ...              ...                ...             ...   
12358523000        399173   00153003210013          E09000007          Camden   
12514325000        397925     TEMP33949894          E09000007          Camden   
13026783000        429519  000760029300001          E09000007          Camden   
13174048000        429164   00153007210028          E09000007          Camden   
13174051000        429169   00153007210029          E09000007          Camden   

                            region                      address_no  \
location_code                                                        
25638063       Inner London - West                              1B   
32905063       Inner London - West                              10   
32907063       Inner London - West                              12   
32907063       Inner London - West                              12   
32912063       Inner London - West                               4   
...                            ...                             ...   
12358523000    Inner London - West         1ST FLR (SOUTH WEST) 32   
12514325000    Inner London - West  GND FLR 7-9 & 1ST FLR LEFT 7-9   
13026783000    Inner London - West                     APARTMENT 8   
13174048000    Inner London - West                      GND FLR 72   
13174051000    Inner London - West                      1ST FLR 72   

                      address_street address_town  pc__pcs  \
location_code                                                
25638063              DOWNSHIRE HILL       LONDON  NW3 1NR   
32905063                  FLASK WALK       LONDON  NW3 1HE   
32907063                  FLASK WALK       LONDON  NW3 1HE   
32907063                  FLASK WALK       LONDON  NW3 1HE   
32912063                  FLASK WALK       LONDON  NW3 1HE   
...                              ...          ...      ...   
12358523000    HAMPSTEAD HIGH STREET       LONDON  NW3 1QD   
12514325000            PERRINS COURT       LONDON  NW3 1QS   
13026783000                  NEW END       LONDON  NW3 1JD   
13174048000    HAMPSTEAD HIGH STREET       LONDON  NW3 1QP   
13174051000    HAMPSTEAD HIGH STREET       LONDON  NW3 1QP   

                                                    use_category  ...  \
location_code                                                     ...   
25638063       Surgeries, Clinics, Health Centres (Rental Val...  ...   
32905063                          Offices (Inc Computer Centres)  ...   
32907063                                                   Shops  ...   
32907063                                                   Shops  ...   
32912063                                                   Shops  ...   
...                                                          ...  ...   
12358523000                       Offices (Inc Computer Centres)  ...   
12514325000                                                Shops  ...   
13026783000                         Showhouses (National Scheme)  ...   
13174048000                                                Shops  ...   
13174051000                       Offices (Inc Computer Centres)  ...   

               duration study-start time-of-entry  time-at-vacant  vacant  \
location_code                                                               
25638063           3524  2010-04-01      494 days            4018       0   
32905063           4018  2010-04-01        0 days            4

In [355]:
subset.to_csv('final-hampstead.csv')

***

In [356]:
## DOUBLE CHECK WE ARE NOT MISSING ANY WHO WERE THERE FROM START TO END (IE. 4019 DAYS)

In [357]:
counter = 0 
stoppingCondition = len(locations)
df2 = pd.DataFrame(data=None, columns=df.columns, index=None)

while counter < stoppingCondition:
        i = locations[counter]
        m = df.loc[i, :].iloc[[ df.loc[i, :]['status_duration'].argmax()]]
        df2 = df2.append(m, ignore_index=False)
        counter += 1

In [358]:
df2

Unnamed: 0       rates_code rates_authority_id rates_authority  \
25638063         25871   00184001B00002          E09000007          Camden   
32905063         24639   00159001009992          E09000007          Camden   
32907063         24727   00159001200006          E09000007          Camden   
32912063         24419   00159000410016          E09000007          Camden   
32945063         24815   00159001500009          E09000007          Camden   
...                ...              ...                ...             ...   
12358523000     399173   00153003210013          E09000007          Camden   
12514325000     397925     TEMP33949894          E09000007          Camden   
13026783000     429519  000760029300001          E09000007          Camden   
13174048000     429164   00153007210028          E09000007          Camden   
13174051000     429169   00153007210029          E09000007          Camden   

                          region                      address_no  \
25638063     Inner London - West                              1B   
32905063     Inner London - West                              10   
32907063     Inner London - West                              12   
32912063     Inner London - West                               4   
32945063     Inner London - West                              15   
...                          ...                             ...   
12358523000  Inner London - West         1ST FLR (SOUTH WEST) 32   
12514325000  Inner London - West  GND FLR 7-9 & 1ST FLR LEFT 7-9   
13026783000  Inner London - West                     APARTMENT 8   
13174048000  Inner London - West                      GND FLR 72   
13174051000  Inner London - West                      1ST FLR 72   

                    address_street address_town  pc__pcs  \
25638063            DOWNSHIRE HILL       LONDON  NW3 1NR   
32905063                FLASK WALK       LONDON  NW3 1HE   
32907063                FLASK WALK       LONDON  NW3 1HE   
32912063                FLASK WALK       LONDON  NW3 1HE   
32945063                FLASK WALK       LONDON  NW3 1HJ   
...                            ...          ...      ...   
12358523000  HAMPSTEAD HIGH STREET       LONDON  NW3 1QD   
12514325000          PERRINS COURT       LONDON  NW3 1QS   
13026783000                NEW END       LONDON  NW3 1JD   
13174048000  HAMPSTEAD HIGH STREET       LONDON  NW3 1QP   
13174051000  HAMPSTEAD HIGH STREET       LONDON  NW3 1QP   

                                                  use_category  ... duration  \
25638063     Surgeries, Clinics, Health Centres (Rental Val...  ...     3524   
32905063                        Offices (Inc Computer Centres)  ...     4018   
32907063                                                 Shops  ...     2192   
32912063                                                 Shops  ...     2722   
32945063                                                 Shops  ...     1461   
...                                                        ...  ...      ...   
12358523000                     Offices (Inc Computer Centres)  ...     1461   
12514325000                                              Shops  ...     1461   
13026783000                       Showhouses (National Scheme)  ...      134   
13174048000                                              Shops  ...      389   
13174051000                     Offices (Inc Computer Centres)  ...      389   

            study-start time-of-entry  time-at-vacant  vacant survival entry  \
25638063     2010-04-01      494 days            4018       0     3524   494   
32905063     2010-04-01        0 days            4018       0     4018     0   
32907063     2010-04-01     1826 days            4018       0     2192  1826   
32912063     2010-04-01     1296 days            4018       0     2722  1296   
32945063     2010-04-01     2557 days            4018       0     1461  2557   
...                 ...           ...             ...     ...      ...   ...   
12358523000  201

In [359]:
df3 = subset.append(df2)

In [360]:
df3

Unnamed: 0       rates_code rates_authority_id rates_authority  \
25638063         25871   00184001B00002          E09000007          Camden   
32905063         24639   00159001009992          E09000007          Camden   
32907063         24701   00159001200006          E09000007          Camden   
32907063         24727   00159001200006          E09000007          Camden   
32912063         24389   00159000410016          E09000007          Camden   
...                ...              ...                ...             ...   
12358523000     399173   00153003210013          E09000007          Camden   
12514325000     397925     TEMP33949894          E09000007          Camden   
13026783000     429519  000760029300001          E09000007          Camden   
13174048000     429164   00153007210028          E09000007          Camden   
13174051000     429169   00153007210029          E09000007          Camden   

                          region                      address_no  \
25638063     Inner London - West                              1B   
32905063     Inner London - West                              10   
32907063     Inner London - West                              12   
32907063     Inner London - West                              12   
32912063     Inner London - West                               4   
...                          ...                             ...   
12358523000  Inner London - West         1ST FLR (SOUTH WEST) 32   
12514325000  Inner London - West  GND FLR 7-9 & 1ST FLR LEFT 7-9   
13026783000  Inner London - West                     APARTMENT 8   
13174048000  Inner London - West                      GND FLR 72   
13174051000  Inner London - West                      1ST FLR 72   

                    address_street address_town  pc__pcs  \
25638063            DOWNSHIRE HILL       LONDON  NW3 1NR   
32905063                FLASK WALK       LONDON  NW3 1HE   
32907063                FLASK WALK       LONDON  NW3 1HE   
32907063                FLASK WALK       LONDON  NW3 1HE   
32912063                FLASK WALK       LONDON  NW3 1HE   
...                            ...          ...      ...   
12358523000  HAMPSTEAD HIGH STREET       LONDON  NW3 1QD   
12514325000          PERRINS COURT       LONDON  NW3 1QS   
13026783000                NEW END       LONDON  NW3 1JD   
13174048000  HAMPSTEAD HIGH STREET       LONDON  NW3 1QP   
13174051000  HAMPSTEAD HIGH STREET       LONDON  NW3 1QP   

                                                  use_category  ... duration  \
25638063     Surgeries, Clinics, Health Centres (Rental Val...  ...     3524   
32905063                        Offices (Inc Computer Centres)  ...     4018   
32907063                                                 Shops  ...     1644   
32907063                                                 Shops  ...     2192   
32912063                                                 Shops  ...     1279   
...                                                        ...  ...      ...   
12358523000                     Offices (Inc Computer Centres)  ...     1461   
12514325000                                              Shops  ...     1461   
13026783000                       Showhouses (National Scheme)  ...      134   
13174048000                                              Shops  ...      389   
13174051000                     Offices (Inc Computer Centres)  ...      389   

            study-start time-of-entry  time-at-vacant  vacant survival entry  \
25638063     2010-04-01      494 days            4018       0     3524   494   
32905063     2010-04-01        0 days            4018       0     4018     0   
32907063     2010-04-01        0 days            1644       0     1644     0   
32907063     2010-04-01     1826 days            4018       0     2192  1826   
32912063     2010-04-01        0 days            1279       0     1279     0   
...                 ...           ...             ...     ...      ...   ...   
12358523000  201

In [361]:
df3 = df3.drop_duplicates()

In [362]:
df3

Unnamed: 0      rates_code rates_authority_id rates_authority  \
25638063         25871  00184001B00002          E09000007          Camden   
32905063         24639  00159001009992          E09000007          Camden   
32907063         24701  00159001200006          E09000007          Camden   
32907063         24727  00159001200006          E09000007          Camden   
32912063         24389  00159000410016          E09000007          Camden   
...                ...             ...                ...             ...   
7993663000      319145  00160000600007          E09000007          Camden   
8773536000      357261  99121006800016          E09000007          Camden   
11040907000     386828  99121001500033          E09000007          Camden   
11040913000     386804  99121001500022          E09000007          Camden   
11765732000     419532  0029500482000A          E09000007          Camden   

                          region        address_no   address_street  \
25638063     Inner London - West                1B   DOWNSHIRE HILL   
32905063     Inner London - West                10       FLASK WALK   
32907063     Inner London - West                12       FLASK WALK   
32907063     Inner London - West                12       FLASK WALK   
32912063     Inner London - West                 4       FLASK WALK   
...                          ...               ...              ...   
7993663000   Inner London - West         STORES AT  STREATLEY PLACE   
8773536000   Inner London - West  BST & GRD FLR 68     HEATH STREET   
11040907000  Inner London - West         BST 15-17     HEATH STREET   
11040913000  Inner London - West     GND FLR 15-17     HEATH STREET   
11765732000  Inner London - West                48     ROSSLYN HILL   

            address_town  pc__pcs  \
25638063          LONDON  NW3 1NR   
32905063          LONDON  NW3 1HE   
32907063          LONDON  NW3 1HE   
32907063          LONDON  NW3 1HE   
32912063          LONDON  NW3 1HE   
...                  ...      ...   
7993663000        LONDON  NW3 1HP   
8773536000        LONDON  NW3 1DN   
11040907000       LONDON  NW3 6TR   
11040913000       LONDON  NW3 6TR   
11765732000       LONDON  NW3 1NH   

                                                  use_category  ... duration  \
25638063     Surgeries, Clinics, Health Centres (Rental Val...  ...     3524   
32905063                        Offices (Inc Computer Centres)  ...     4018   
32907063                                                 Shops  ...     1644   
32907063                                                 Shops  ...     2192   
32912063                                                 Shops  ...     1279   
...                                                        ...  ...      ...   
7993663000                                              Stores  ...     2557   
8773536000                                               Shops  ...      921   
11040907000                     Offices (Inc Computer Centres)  ...     2192   
11040913000                                              Shops  ...     2192   
11765732000                                        Restaurants  ...      625   

            study-start time-of-entry  time-at-vacant  vacant survival entry  \
25638063     2010-04-01      494 days            4018       0     3524   494   
32905063     2010-04-01        0 days            4018       0     4018     0   
32907063     2010-04-01        0 days            1644       0     1644     0   
32907063     2010-04-01     1826 days            4018       0     2192  1826   
32912063     2010-04-01        0 days            1279       0     1279     0   
...                 ...           ...             ...     ...      ...   ...   
7993663000   2010-04-01     1461 days            4018       0     2557  1461   
8773536000   2010-04-01     1454 days            2375       0      921  1454   
11040907000  2010-04-01     1826 days            4018       0     2192  1826   
11040913000  2010-04-01     

In [363]:
## if time-at-vacant = 4018 then Vacant change to 1 (ie it died)

In [364]:
df3.loc[df3['time-at-vacant'] < 4018, ['vacant']] = 1

/opt/conda/envs/sds2020/lib/python3.7/site-packages/pandas/core/indexing.py:1765: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  isetter(loc, value)


In [365]:
df3

Unnamed: 0      rates_code rates_authority_id rates_authority  \
25638063         25871  00184001B00002          E09000007          Camden   
32905063         24639  00159001009992          E09000007          Camden   
32907063         24701  00159001200006          E09000007          Camden   
32907063         24727  00159001200006          E09000007          Camden   
32912063         24389  00159000410016          E09000007          Camden   
...                ...             ...                ...             ...   
7993663000      319145  00160000600007          E09000007          Camden   
8773536000      357261  99121006800016          E09000007          Camden   
11040907000     386828  99121001500033          E09000007          Camden   
11040913000     386804  99121001500022          E09000007          Camden   
11765732000     419532  0029500482000A          E09000007          Camden   

                          region        address_no   address_street  \
25638063     Inner London - West                1B   DOWNSHIRE HILL   
32905063     Inner London - West                10       FLASK WALK   
32907063     Inner London - West                12       FLASK WALK   
32907063     Inner London - West                12       FLASK WALK   
32912063     Inner London - West                 4       FLASK WALK   
...                          ...               ...              ...   
7993663000   Inner London - West         STORES AT  STREATLEY PLACE   
8773536000   Inner London - West  BST & GRD FLR 68     HEATH STREET   
11040907000  Inner London - West         BST 15-17     HEATH STREET   
11040913000  Inner London - West     GND FLR 15-17     HEATH STREET   
11765732000  Inner London - West                48     ROSSLYN HILL   

            address_town  pc__pcs  \
25638063          LONDON  NW3 1NR   
32905063          LONDON  NW3 1HE   
32907063          LONDON  NW3 1HE   
32907063          LONDON  NW3 1HE   
32912063          LONDON  NW3 1HE   
...                  ...      ...   
7993663000        LONDON  NW3 1HP   
8773536000        LONDON  NW3 1DN   
11040907000       LONDON  NW3 6TR   
11040913000       LONDON  NW3 6TR   
11765732000       LONDON  NW3 1NH   

                                                  use_category  ... duration  \
25638063     Surgeries, Clinics, Health Centres (Rental Val...  ...     3524   
32905063                        Offices (Inc Computer Centres)  ...     4018   
32907063                                                 Shops  ...     1644   
32907063                                                 Shops  ...     2192   
32912063                                                 Shops  ...     1279   
...                                                        ...  ...      ...   
7993663000                                              Stores  ...     2557   
8773536000                                               Shops  ...      921   
11040907000                     Offices (Inc Computer Centres)  ...     2192   
11040913000                                              Shops  ...     2192   
11765732000                                        Restaurants  ...      625   

            study-start time-of-entry  time-at-vacant  vacant survival entry  \
25638063     2010-04-01      494 days            4018       0     3524   494   
32905063     2010-04-01        0 days            4018       0     4018     0   
32907063     2010-04-01        0 days            1644       1     1644     0   
32907063     2010-04-01     1826 days            4018       0     2192  1826   
32912063     2010-04-01        0 days            1279       1     1279     0   
...                 ...           ...             ...     ...      ...   ...   
7993663000   2010-04-01     1461 days            4018       0     2557  1461   
8773536000   2010-04-01     1454 days            2375       1      921  1454   
11040907000  2010-04-01     1826 days            4018       0     2192  1826   
11040913000  2010-04-01     

In [366]:
df3.to_csv('final.csv')